src/index_map.rs

```diff
@@ -0,0 +1,432 @@
+//! [`NEIndexMap`] is a non-empty variant of [`IndexMap`].
+#![deny(clippy::correctness, clippy::suspicious)]
+#![warn(clippy::complexity, clippy::perf, clippy::style, clippy::pedantic)]
+#![allow(clippy::module_name_repetitions)]
```

Comment:
Can you write up a quick justification for these?

Reply:
The first two lines enable the named [clippy lint groups](https://rust-lang.github.io/rust-clippy/stable/index.html). Ideally these lints are enabled for the entire project in the `Cargo.toml` and enforced in CI but since this is out of scope for this PR and I didn't want to miss out on the help of Clippy, I decided to selectively enable them in this file only.

Of the selected lint groups, `pedantic` is probably the most controversial. In fact, the [documentation](https://doc.rust-lang.org/stable/clippy/lints.html#pedantic) recommends to use `#[allow]` statements where needed. 

And this brings us to the third line above, the [`module_name_repetitions` lint](https://rust-lang.github.io/rust-clippy/stable/index.html#/module_name_repetitions?groups=pedantic) disallows name repetitions such as `mod indexmap { struct IndexMap; }`, I enabled this lint here because it seems appropriate.